### PR TITLE
Support extraction by filename as well as directory

### PR DESCRIPTION
**Pull request by ENuge**
Reason for the change: https://github.com/python-babel/babel/issues/253 . This obviates the need for parallelizing to make things faster (on the consumer's side, we only scan files that are different in a given branch compared to origin's master, which brings scan time down from ~4min 30s to ~5-20s, depending on how old one's branch is). 

Code changes: moved half of the logic from extract_from_dir into its own function. Then added a few "is this a path or directory?" type checks to frontend.py to use the right thing. Those checks are kind of ugly right now but I couldn't think of a more elegant way of doing it. Added/rejiggered a couple of tests.

### CODE REVIEW

**Comment by sils1297** [comment]
Hi, can you fix the issues pointed out by the gitmate bot and refactor your stuff into atomic commits as described in http://coala.readthedocs.org/en/latest/Getting_Involved/Writing_Good_Commits/ ?

**Comment by sils1297** [comment]
Thanks for submitting :)

**Comment by ENuge** [comment]
Appeased the LineLengthBot with Yelp's preferred syntax (which is vertically long, but eh this line is super-awkward regardless).

I'll refactor my commits later today or tomorrow. (I started this change on Yelp's old fork of babel, so this version was more of a copy/paste job. But I'll make the history make some sense.)

**Comment by sils1297** [comment]
@ENuge cool, ping us when it's done.

File: babel/messages/frontend.py [hunk 1]
```diff
@@ -333,29 +337,48 @@ def run(self):
...
+                if os.path.isdir(path):
+                    extracted = extract_from_dir(
+                        path, method_map, options_map,
+                        keywords=self._keywords,
+                        comment_tags=self.add_comments,
+                        callback=callback,
+                        strip_comment_tags=self.strip_comments
+                    )
+                else:
+                    extracted = check_and_call_extract_file(path, method_map,
```
Comment (akx) [inline babel/messages/frontend.py]: Could you use the same indent style as in the `extract_from_dir` call above? :)

File: babel/messages/frontend.py [hunk 2]
```diff
@@ -333,29 +337,48 @@ def run(self):
...
+                                                            options_map, callback, self._keywords,
+                                                            self.add_comments, self.strip_comments)
                 for filename, lineno, message, comments, context in extracted:
-                    filepath = os.path.normpath(os.path.join(dirname, filename))
+                    if os.path.isdir(path):
+                        filepath = os.path.normpath(os.path.join(path, filename))
+                    else:
+                        filepath = filename
```
Comment (akx) [inline babel/messages/frontend.py]: Add a comment like `  # already normalized`

File: babel/messages/extract.py [hunk 3]
```diff
@@ -142,28 +142,74 @@ def extract_from_dir(dirname=None, method_map=DEFAULT_MAPPING,
...
+
+    for pattern, method in method_map:
+        if pathmatch(pattern, filename):
+            options = {}
+            for opattern, odict in options_map.items():
+                if pathmatch(opattern, filename):
+                    options = odict
+            if callback:
+                callback(filename, method, options)
+            for lineno, message, comments, context in \
```
Comment (akx) [inline babel/messages/extract.py]: Maybe don't extract the tuple here?

```python
for message_tuple in extract...:
   yield (filename,) + message_tuple
```

File: babel/messages/extract.py [hunk 4]
```diff
@@ -142,28 +142,74 @@ def extract_from_dir(dirname=None, method_map=DEFAULT_MAPPING,
...
+    :param dirpath: the path to the directory to extract messages from.
+    """
+    if dirpath is None:
+        dirpath = os.getcwd()
+
+    # filename is the relative path from dirpath to the actual file
+    filename = relpath(filepath, dirpath)
+
+    for pattern, method in method_map:
+        if pathmatch(pattern, filename):
```
Comment (akx) [inline babel/messages/extract.py]: Would be better to invert this if (avoids over-indentation):

```python
if not pathmatch(pattern, filename):
  continue
options = ....
```

File: babel/messages/extract.py [hunk 5]
```diff
@@ -142,28 +142,74 @@ def extract_from_dir(dirname=None, method_map=DEFAULT_MAPPING,
...
+                     tuples that specify which of their arguments contain
+                     localizable strings
+    :param comment_tags: a list of tags of translator comments to search for
+                         and include in the results
+    :param strip_comment_tags: a flag that if set to `True` causes all comment
+                               tags to be removed from the collected comments.
+    :param dirpath: the path to the directory to extract messages from.
+    """
+    if dirpath is None:
+        dirpath = os.getcwd()
```
Comment (akx) [inline babel/messages/extract.py]: This should not be done in this function; it's fine to just do it in the caller.

File: babel/messages/extract.py [hunk 6]
```diff
@@ -142,28 +142,74 @@ def extract_from_dir(dirname=None, method_map=DEFAULT_MAPPING,
...
+                     positional arguments, in that order
+    :param keywords: a dictionary mapping keywords (i.e. names of functions
+                     that should be recognized as translation functions) to
+                     tuples that specify which of their arguments contain
+                     localizable strings
+    :param comment_tags: a list of tags of translator comments to search for
+                         and include in the results
+    :param strip_comment_tags: a flag that if set to `True` causes all comment
+                               tags to be removed from the collected comments.
+    :param dirpath: the path to the directory to extract messages from.
```
Comment (akx) [inline babel/messages/extract.py]: Can you expand a little on this -- namely that this is used to calculate the relative names of files?

File: babel/messages/frontend.py [hunk 7]
```diff
@@ -300,17 +300,21 @@ def finalize_options(self):
...
             ]).keys()
 
-        if not self.input_dirs:
-            raise DistutilsOptionError("no input directories specified")
+        if not self.input_paths:
+            raise DistutilsOptionError("no input files or directories specified")
+
+        for path in self.input_paths:
+            if not os.path.isdir(path) and not os.path.isfile(path):
+                raise DistutilsOptionError("Input path: %s is not a file or directory" % path)
```
Comment (akx) [inline babel/messages/frontend.py]: Is this even possible? :D

File: babel/messages/extract.py [hunk 8]
```diff
@@ -142,28 +142,74 @@ def extract_from_dir(dirname=None, method_map=DEFAULT_MAPPING,
...
-                                            keywords=keywords,
-                                            comment_tags=comment_tags,
-                                            options=options,
-                                            strip_comment_tags=
-                                                strip_comment_tags):
-                        yield filename, lineno, message, comments, context
-                    break
+            filepath = os.path.join(root, filename).replace(os.sep, '/')
+
+            for (
```
Comment (akx) [inline babel/messages/extract.py]: Don't expand the tuple here either.

```python
for message_tuple in check...:
    yield message_tuple
```

is fine and dandy!

**Comment by akx** [comment]
And aside from the handful of comments I had, great work @ENuge ! :)

I'd appreciate a test that passes in a direct filename to the command line command though. :)

File: babel/messages/frontend.py (hunk 7 above)
Comment (ENuge) [inline babel/messages/frontend.py]: Heh, this is meant to be checking if the path exists. I didn't really think about the much more direct way of doing that...namely, using os.path.exists(..). Will update.

File: babel/messages/extract.py (hunk 8 above)
Comment (ENuge) [inline babel/messages/extract.py]: Cool, I'll add a comment/docstring line with the return type or something to make up for the loss in explicitness.

File: babel/messages/frontend.py (hunk 1 above)
Comment (ENuge) [inline babel/messages/frontend.py]: Yup! Done.

File: babel/messages/extract.py (hunk 5 above)
Comment (ENuge) [inline babel/messages/extract.py]: fwiw extract_from_dir(..) above it does the same check. Though I'm fine with having the caller explicitly pass it in.

**Comment by akx** [comment]
@ENuge: Looking good!

Please squash all of the commits into one though :)

**Comment by sils1297** [comment]
Interesting, travis errored out on mac only for pypy 2.6, maybe a nondeterministic bug?

**Comment by akx** [comment]
@sils1297 Heh, yes... race condition against the clock, it seems. If the wallclock ticks forward a second while those tests are being run, they may fail.

[Should probably freeze time for the duration of those tests.](https://pypi.python.org/pypi/freezegun)

**Comment by akx** [comment]
Super nice work @ENuge!

I'm gonna merge this, but come to think of it, we should add an alias for `input-paths` called `input-dirs` for backward compatibility. I'll make an issue out of that. :)

**Pull request merged**